### DISCUSSION TITLE
Zustand not rerendering

### DISCUSSION THREAD

**Discussion by Gopikrishna-KS** [pmndrs/zustand#2383] [Q&A]
I am a beginner to Zustand. I have a simple use case where there are parent, child components. In the child component, I am using Zustand to update the state, it is getting logged but not re-rendering, could someone please help?

This is the related sandbox - [link](https://codesandbox.io/p/sandbox/inspiring-mountain-z93fi?file=%2Fsrc%2FuseTodos.ts&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clta1kf2h0007356h7i9507on%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clta1kf2g0003356huhqohmbs%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clta1kf2g0004356hgfectfuc%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clta1kf2g0006356hwed1ipkl%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clta1kf2g0003356huhqohmbs%2522%253A%257B%2522id%2522%253A%2522clta1kf2g0003356huhqohmbs%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clta1kf2g0006356hwed1ipkl%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clta1kf2g0005356h5txkpg1e%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clta1kf2g0006356hwed1ipkl%2522%252C%2522activeTabId%2522%253A%2522clta1kf2g0005356h5txkpg1e%2522%257D%252C%2522clta1kf2g0004356hgfectfuc%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clta1kf2g0004356hgfectfuc%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Edit: Not sure why sandbox is not loading, I have pasted the code below.

Child.tsx
```
import { shallow } from "zustand/shallow";
import useStore, { testType } from "./store";

const selector = (state: testType) => ({
  type1: state.type1,
  setValue: state.setValue,
});

export default () => {
  const { type1, setValue } = useStore(selector, shallow);

  return (
    <input
      value={type1.get("0")!.get("label")}
      onChange={(e) => setValue("0", "label", e.target.value)}
    />
  );
};
```

App.tsx

```
import Child from "./Child";

export default function Home() {
  return <Child />;
}
```

store.ts

```
import create from "zustand";

export type testType = {
  type1: Map<string, Map<string, any>>;
  setValue: (key: string, mapKey: string, mapValue: any) => void;
};

const testMap = new Map([["label", "testValue"]]);
const testMap2 = new Map([["0", testMap]]);

const useStore = create<testType>((set, get) => ({
  type1: testMap2,
  setValue: (key: string, mapKey: string, mapValue: any) => {
    const tempType1 = get().type1;
    const specificType1 = tempType1.get(key);
    specificType1!.set(mapKey, mapValue);
    tempType1.set(key, specificType1!);
    console.log(tempType1);
    set({
      type1: tempType1,
    });
  },
}));

export default useStore;
```

**Comment by dbritto-dev** [comment]
@Gopikrishna-KS you should keep in mind that immutability is important in react, when you mutate an object directly couldn't trigger any update/render
